### PR TITLE
An implementation of backtracking solver

### PR DESCRIPTION
**Pull request by horenmar**
This adds a fairly minimal implementation of a backtracking lock-chart solver, together with its dependencies and interfaces. The parts are

* An interface for key cutting generators.
  * Two implementations of this interface, one that performs an exhaustive check over the cutting space, and one that uses MiniSat to repeatedly solve a 1x0 lock-chart for the same geometry.
* A class for caching cuttings generated by these generators
  * Important for the backtracking, as multi-profile lock-charts mean that we want to be able to reuse the same cutting for multiple keys.
* A non-recursive backtracking solver.
  * The commit history also intentionally contains a simple recursive implementation, as it might be needed at some point for testing purposes.

_A CLI implementation will be another PR, pending #30_

### CODE REVIEW

File: src/main/cpp/backtracking-solver.cpp [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+#include <cassert>
+#include <stack>
+
+
+namespace mks {
+
+namespace {
+
+enum class state {
```
Comment (cernoch) [inline src/main/cpp/backtracking-solver.cpp]: A few more comments would be helpful.

Overall, the code is quite easy to read. Just this part was the hardest for me.

File: src/main/cpp/backtracking-solver.cpp [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+        stack.push({ 0, 0 });
+
+        while (!stack.empty()) {
+            auto cstate = stack.top(); stack.pop();
+            assert(cstate.key() <= m_chart.keys() &&
+                   "Attempt to solve for non-sensical key");
+
+            // Key 1-after means we have found solution for all existing keys
+            if (cstate.key() == m_chart.keys()) {
+                m_solution.validate(m_chart, true);
```
Comment (cernoch) [inline src/main/cpp/backtracking-solver.cpp]: Why is this line here? The `mks::solution::validate(...)` returns true/false, it does not throw any exception. AFAIK, this line should have no side effect.

File: src/main/cpp/backtracking-solver.cpp [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+
+
+            if (!cstate.is_fresh()) {
+                ++keys_reassigned;
+                m_solution.unassign(cstate.key(), m_chart);
+                cstate.set_idx(m_holder.next_index(cstate.idx()));
+            }
+
+            if (cstate.is_fresh()) {
+                cstate.open();
```
Comment (cernoch) [inline src/main/cpp/backtracking-solver.cpp]: I'm just curious. What's the difference between `state::fresh` and `state::open`? Can you merge them?

File: src/main/cpp/backtracking-solver.cpp [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+                    ++keys_assigned;
+                    stack.push(cstate);
+                    stack.push(cstate.make_succ());
+                    break;
+                } else {
+                    m_solution.unassign(cstate.key(), m_chart);
+                    cstate.set_idx(m_holder.next_index(cstate.idx()));
+                }
+            }
+        }
```
Comment (cernoch) [inline src/main/cpp/backtracking-solver.cpp]: The `while` loop above is such a pleasure to read! Really, it's simple, clean, easy to understand. I really like it. 👍

File: src/main/cpp/cutting-generator-sat.cpp [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    m_stp.add_key(0);
+    for (auto const& gecon : rhs.m_stp.gecons_) {
+        m_stp.add_gecon(gecon);
+    }
+}
+
+cutting sat_cutting_generator::next() {
+    if (m_used_up) { return {}; }
+
+    auto succ = m_stp.solve();
```
Comment (cernoch) [inline src/main/cpp/cutting-generator-sat.cpp]: This line will soon drain the performance. For `positions <= 4` and `depths <= 4`, this call will be really quick. For values above 5, or 6 this may take too long.

As a quick fix, let's add value 100 as a conflict limit:
```c++
auto succ = m_stp.solve(100);
```
Or make this a parameter...

File: src/main/cpp/cutting-holder.hpp [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+     * Returns next valid index into the holder.
+     *
+     * Generally this means that index increases, but if the generator
+     * stops returning extra shapes, then it can also wrap around to 0.
+     *
+     * If the provided index is not valid, then the behaviour is undefined.
+     *
+     * @param[in] current_idx a currently valid index
+     */
+    size_t next_index(size_t current_idx);
```
Comment (cernoch) [inline src/main/cpp/cutting-holder.hpp]: This API function is quite enigmatic. 😉 However, documentation is good; hence let's keep the code as it is.

File: src/main/cpp/backtracking-solver.cpp [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+        return m_state == state::fresh;
+    }
+
+    void open() {
+        m_state = state::open;
+    }
+
+private:
+    key_id m_key;
+    size_t m_idx;
```
Comment (cernoch) [inline src/main/cpp/backtracking-solver.cpp]: Could you provide a comment, which links this to `cutting_holder::next_index`, please? That's should be enough to make this class easy to read.

File: src/main/cpp/backtracking-solver.hpp [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+/**
+ * A simple backtracking solver.
+ *
+ * In the future it will need to become smarter, but at this moment
+ * it works by assigning a generated cutting to every key in chart,
+ * and backtracking if the assignment is invalid.
+ */
+class backtracking_solver {
+
+    geometry m_geom;
```
Comment (cernoch) [inline src/main/cpp/backtracking-solver.hpp]: The generators have a similar field declared as
```c++
geometry const& m_geom;
```
Therefore, when using a `backtracking_solver`, the `geometry` must live externally anyway.

I can think of a few simple solutions:
1) Make this line a const reference as well.
2) Remove the `m_geom` field from all generators and let it be passed as a parameter in all functions (similarly as in `mks::solution`). **Main benefit:** the `backtracking_solver` will be self-contained, easy to move, easy to keep in a `vector`, ...
3) Add a templated constructor, which initializes `m_holder` as `std::make_unique<TEMPLATE>(m_geom)`. Of course, copying and moving the solver should be prohibited to keep the references valid. **Main benefit:** Simplification of the client code (the geometry will not have to live at the client).

**Comment by cernoch** [comment]
Not sure about bb3628e.

My feeling is as follows: Since MiniSat (and PBlib, added in future) are implemented in C++11, maybe we should stay in this world.

Is it a bad idea? If so, please feel free to reject the patch.

File: src/main/cpp/backtracking-solver.cpp (hunk 2 above)
Comment (horenmar) [reply src/main/cpp/backtracking-solver.cpp]: That shouldn't be there, it is a leftover debugging artifact.

File: src/main/cpp/backtracking-solver.cpp (hunk 3 above)
Comment (horenmar) [reply src/main/cpp/backtracking-solver.cpp]: The difference is simple:
State is fresh when it is new.
State is open when it has been taken out of stack at least once before **AND** it has not been moved yet.

The difference is that a fresh state can be used as-is by the cutting-assigning loop as is, while non-fresh states need to be moved before their index is used.

`state::open` and `state::moved` need to be disambiguated to determine whether the current index could've wrapped around (and thus the state can be closed), or it has not (and while the current index is equal with original index, the state is not closed yet).

File: src/main/cpp/cutting-generator-sat.cpp (hunk 5 above)
Comment (horenmar) [reply src/main/cpp/cutting-generator-sat.cpp]: I'll add a defaulted conflict limit as a constructor parameter to `sat_cutting_generator`.

**Comment by horenmar** [comment]
Oooof, a bad rebase auto-closed the pull request. GitHub ¯\\\_(ツ)_/¯

**Comment by cernoch** [comment]
Maybe GitHub's opened my inbox, read the e-mail addressed to you, understood it using a very sophisticated AI and decided that the PR is about to be closed anyway. 😄
